### DISCUSSION TITLE
Properly rendering rich text even if start tags are no longer visible

### DISCUSSION THREAD

**Discussion by iiPythonx** [Textualize/rich#1353] [Q&A] [answered]
I'm working on a in-terminal text viewer of sorts. It displays text using the `rich.panel.Panel` class, and the text is only what the user would be able to see (based on their line number, how big their terminal is, etc). I have a problem though. If you scroll down in a file, you can occasionally run into rich formatting errors since certain tags get cut off since when they are no longer printed.

My question, what is the **best** way of allowing scrolling while also not causing formatting errors.
I have attempted using regex to find all cut off tags and reapply them at the beginning of the panels text, which works but I feel like it's kind of hacky.

Any help would be appreciated. I know it's a weird question but I seriously feel like I'm doing it wrong.

**Comment by willmcgugan** [comment, marked as answer]
You probably don't want to do basic string operations on console markup if there is a chance you can break tags.

You can convert console markup in to a [Text](https://rich.readthedocs.io/en/latest/reference/text.html) instance, and then further process it. 

Here's how you can generate a Text object from markup:

```python
text = Text.from_markup(my_markup)
```

**Reply by iiPythonx** [reply]
I'll give that a try, probably attempt to cut out anything that wouldn't be rendered.
Thanks for your time.
